storage: TestNodeLivenessStatusMap failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/07edde23460e8ffe0ec40f89975c3b95fc28343e

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestNodeLivenessStatusMap PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1030855&tab=buildLog

```
I181129 07:03:56.660099 114222 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n2: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:42207: connect: connection refused"
I181129 07:03:56.663304 113051 sql/event_log.go:126  [in test] Event: "node_decommissioned", target: 4, info: {}
I181129 07:03:56.663345 113051 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181129 07:03:56.663603 113051 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
I181129 07:03:56.668643 114223 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n4: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: <nil>
I181129 07:03:56.669041 113051 storage/node_liveness_test.go:865  [in test] done removing node 4
I181129 07:03:56.669072 113051 storage/node_liveness_test.go:867  [in test] checking status map
--- FAIL: TestNodeLivenessStatusMap (0.54s)
=== RUN   TestNodeLivenessStatusMap/DEAD
=== PAUSE TestNodeLivenessStatusMap/DEAD
=== CONT  TestNodeLivenessStatusMap/DEAD
    --- FAIL: TestNodeLivenessStatusMap/DEAD (45.09s)
    	node_liveness_test.go:888: condition failed to evaluate within 45s: unexpected status: got UNAVAILABLE, expected DEAD
    		goroutine 114173 [running]:
    		runtime/debug.Stack(0xa7a358200, 0xc42097e5a0, 0x2dddec0)
    			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
    		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x2e3c1c0, 0xc4210285a0, 0xc42097e540)
    			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x10d
    		github.com/cockroachdb/cockroach/pkg/storage_test.TestNodeLivenessStatusMap.func1(0xc4210285a0)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/node_liveness_test.go:888 +0xcc
    		testing.tRunner(0xc4210285a0, 0xc4210e2540)
    			/usr/local/go/src/testing/testing.go:777 +0xd0
    		created by testing.(*T).Run
    			/usr/local/go/src/testing/testing.go:824 +0x2e0
=== RUN   TestNodeLivenessStatusMap/DECOMMISSIONED
=== PAUSE TestNodeLivenessStatusMap/DECOMMISSIONED
I181129 07:03:56.669234 114176 util/stop/stopper.go:537  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181129 07:03:56.669249 114178 util/stop/stopper.go:537  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181129 07:03:56.670193 114176 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
I181129 07:03:56.670258 114178 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181129 07:03:56.670367 114178 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
=== CONT  TestNodeLivenessStatusMap/DECOMMISSIONED
I181129 07:03:56.724871 114172 storage/node_liveness_test.go:889  [in test] checking expected status for node 1
    --- FAIL: TestNodeLivenessStatusMap/DECOMMISSIONED (45.09s)
    	node_liveness_test.go:888: condition failed to evaluate within 45s: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
    		goroutine 114175 [running]:
    		runtime/debug.Stack(0xa7a358200, 0xc42089b3e0, 0x2dddec0)
    			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
    		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x2e3c1c0, 0xc421028780, 0xc42089af30)
    			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x10d
    		github.com/cockroachdb/cockroach/pkg/storage_test.TestNodeLivenessStatusMap.func1(0xc421028780)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/node_liveness_test.go:888 +0xcc
    		testing.tRunner(0xc421028780, 0xc4210e25e0)
    			/usr/local/go/src/testing/testing.go:777 +0xd0
    		created by testing.(*T).Run
    			/usr/local/go/src/testing/testing.go:824 +0x2e0

```

SHA: https://github.com/cockroachdb/cockroach/commits/6fb1b0037ead768572b49646e0ffbe73013a8f9a

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestNodeLivenessStatusMap PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1033100&tab=buildLog

```
I181130 07:06:46.717505 113722 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
I181130 07:06:47.717755 113720 storage/node_liveness_test.go:889  [in test] checking expected status for node 2
I181130 07:06:47.717811 113720 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got UNAVAILABLE, expected DEAD
I181130 07:06:47.717755 113722 storage/node_liveness_test.go:889  [in test] checking expected status for node 4
I181130 07:06:47.717837 113722 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
I181130 07:06:48.718014 113722 storage/node_liveness_test.go:889  [in test] checking expected status for node 4
I181130 07:06:48.718065 113722 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
I181130 07:06:48.718025 113720 storage/node_liveness_test.go:889  [in test] checking expected status for node 2
I181130 07:06:48.718083 113720 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got UNAVAILABLE, expected DEAD
I181130 07:06:49.718296 113720 storage/node_liveness_test.go:889  [in test] checking expected status for node 2
I181130 07:06:49.718317 113722 storage/node_liveness_test.go:889  [in test] checking expected status for node 4
I181130 07:06:49.718358 113720 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got UNAVAILABLE, expected DEAD
I181130 07:06:49.718359 113722 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
I181130 07:06:50.718569 113720 storage/node_liveness_test.go:889  [in test] checking expected status for node 2
I181130 07:06:50.718642 113720 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got UNAVAILABLE, expected DEAD
I181130 07:06:50.718569 113722 storage/node_liveness_test.go:889  [in test] checking expected status for node 4
I181130 07:06:50.718657 113722 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
I181130 07:06:51.718884 113720 storage/node_liveness_test.go:889  [in test] checking expected status for node 2
I181130 07:06:51.718942 113720 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got UNAVAILABLE, expected DEAD
I181130 07:06:51.718885 113722 storage/node_liveness_test.go:889  [in test] checking expected status for node 4
I181130 07:06:51.718975 113722 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
I181130 07:06:52.719127 113722 storage/node_liveness_test.go:889  [in test] checking expected status for node 4
I181130 07:06:52.719187 113722 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
I181130 07:06:52.719127 113720 storage/node_liveness_test.go:889  [in test] checking expected status for node 2
I181130 07:06:52.719215 113720 storage/node_liveness_test.go:888  SucceedsSoon: unexpected status: got UNAVAILABLE, expected DEAD
    --- FAIL: TestNodeLivenessStatusMap/DECOMMISSIONED (45.09s)
    	node_liveness_test.go:888: condition failed to evaluate within 45s: unexpected status: got DECOMMISSIONING, expected DECOMMISSIONED
    		goroutine 113722 [running]:
    		runtime/debug.Stack(0xa7a358200, 0xc4208dade0, 0x2dddf00)
    			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
    		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x2e3c2a0, 0xc4226a8b40, 0xc4208dad50)
    			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x10d
    		github.com/cockroachdb/cockroach/pkg/storage_test.TestNodeLivenessStatusMap.func1(0xc4226a8b40)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/node_liveness_test.go:888 +0xcc
    		testing.tRunner(0xc4226a8b40, 0xc421679400)
    			/usr/local/go/src/testing/testing.go:777 +0xd0
    		created by testing.(*T).Run
    			/usr/local/go/src/testing/testing.go:824 +0x2e0

```

--- a/pkg/storage/node_liveness_test.go
+++ b/pkg/storage/node_liveness_test.go
@@ -832,16 +832,6 @@ func TestNodeLivenessStatusMap(t *testing.T) {
 
 	firstServer := tc.Server(0).(*server.TestServer)
 
-	// Ensure that dead nodes are quickly recognized as dead by gossip.
-	// We set the cluster setting after we've done propagating statuses
-	// above so that the new setting value does not race with gossip
-	// initialization.
-	// Note: morally we'd probably need to use SQL here to set
-	// the cluster setting. However that restricts the allowable values
-	// to 1m15s, which is much too long for a test.
-	storage.TimeUntilStoreDead.Override(&firstServer.ClusterSettings().SV,
-		storage.TestTimeUntilStoreDead)
-
 	liveNodeID := firstServer.NodeID()
 
 	deadNodeID := tc.Server(1).NodeID()
@@ -886,6 +876,16 @@ func TestNodeLivenessStatusMap(t *testing.T) {
 			t.Parallel()
 
 			testutils.SucceedsSoon(t, func() error {
+				// Ensure that dead nodes are quickly recognized as dead by
+				// gossip. Overriding cluster settings is generally a really bad
+				// idea as they are also populated via Gossip and so our update
+				// is possibly going to be wiped out. But going through SQL
+				// doesn't allow durations below 1m15s, which is much too long
+				// for a test.
+				// We do this in every SucceedsSoon attempt, so we'll be good.
+				storage.TimeUntilStoreDead.Override(&firstServer.ClusterSettings().SV,
+					storage.TestTimeUntilStoreDead)
+
 				log.Infof(ctx, "checking expected status for node %d", nodeID)
 				nodeStatuses := callerNodeLiveness.GetLivenessStatusMap()
 				if st, ok := nodeStatuses[nodeID]; !ok {

